### PR TITLE
US003 view role capability

### PR DESCRIPTION
**Pull request by malachi-mccarney**
Capability tables added for future tickets, JobRole swagger endpoint updated to view Capability name linked though the Job Families table

### CODE REVIEW

File: src/main/java/com/kainos/ea/dao/JobRoleDao.java [hunk 1]
```diff
@@ -14,20 +14,24 @@ public class JobRoleDao
     public List<JobRole> getRoles(Connection c) throws SQLException {
         Statement st = c.createStatement();
 
-        ResultSet rs = st.executeQuery(
-                "SELECT role_id, role_title, spec_link"
-                        + " FROM Job_Roles;");
+
+        ResultSet rs = st.executeQuery(" SELECT Job_Roles.role_id, Job_Roles.role_title, Capabilities.name, Job_Roles.spec_link, Job_Roles.spec_summary" +
```
Comment (Patrick-Mc-Grath) [inline src/main/java/com/kainos/ea/dao/JobRoleDao.java:18]: Do we need the spec_link and spec_summary for this?

File: src/main/java/com/kainos/ea/model/JobRole.java [hunk 2]
```diff
@@ -8,16 +8,23 @@ public class JobRole
     private int roleId;
     private String roleTitle;
     private String specLink;
+    private String capabilityName;
```
Comment (Patrick-Mc-Grath) [inline src/main/java/com/kainos/ea/model/JobRole.java:11]: You can create a response class for this instead of changing JobRole itself

**Review by Patrick-Mc-Grath — COMMENTED** [review]
Noticed a few issues, otherwise looks good to me.

File: src/main/java/com/kainos/ea/controller/CapabilitiesController.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.sql.SQLException;
+
+@Api("API Hack Street Boys")
+@Path("/api")
+
+public class CapabilitiesController {
```
Comment (ccurran01) [inline src/main/java/com/kainos/ea/controller/CapabilitiesController.java:19]:
```suggestion
public class CapabilityController {
```

File: src/test/java/com/kainos/ea/service/CapabilityServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.ArrayList;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@ExtendWith(MockitoExtension.class)
+class CapabilityServiceTest {
```
Comment (ccurran01) [inline src/test/java/com/kainos/ea/service/CapabilityServiceTest.java:20]: Looks like we are missing a test for when the database connector throws a connection exception

File: src/test/java/com/kainos/ea/controller/CapabilityControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+class CapabilityControllerTest {
+
+    CapabilitiesService capabilitiesService = Mockito.mock(CapabilitiesService.class);
+
+    CapabilitiesController capabilitiesController = new CapabilitiesController(capabilitiesService);
+
+    static final DropwizardAppExtension<WebServiceConfiguration> APP = new DropwizardAppExtension<>(
+            WebServiceApplication.class, null,
+            new ResourceConfigurationSourceProvider()
+    );
```
Comment (ccurran01) [inline src/test/java/com/kainos/ea/controller/CapabilityControllerTest.java:28]: Don't think that we need this

**Review by ccurran01 — COMMENTED** [review]
Think there's some confusion about how capabilities and roles and families are related. 

A role belongs to a capability 
a capability belongs to a job family.

File: SQLScripts/US003-view-job-role-capability.sql [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+	description varchar(100) NOT NULL
+
+);
+
+ALTER TABLE Job_Roles
+ADD COLUMN capability_id  smallint;
+
+ALTER TABLE Job_Roles
+ADD CONSTRAINT FK1_capability_id
+FOREIGN KEY (capability_id) REFERENCES Capabilities(capability_id);
```
Comment (ccurran01) [inline SQLScripts/US003-view-job-role-capability.sql:13]: Could these statements not be merged?

File: src/main/java/com/kainos/ea/controller/CapabilityController.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+    @GET
+    @Path("/capabilities")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getCapabilities()
+    {
+        try
+        {
+            return Response.ok(capabilitiesService.getCapabilities()).build();
+        } catch (SQLException | DatabaseConnectionException e)
```
Comment (ccurran01) [inline src/main/java/com/kainos/ea/controller/CapabilityController.java:38]: Could we create a custom exception that we can use for this scenario? 

```suggestion
        } catch (FailedToGetCapabilityException e)
```

File: src/main/java/com/kainos/ea/model/Capabilities.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
+package com.kainos.ea.model;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class Capabilities {
+    private int CapabilityId;
+    private String Name;
+    private String Description;
```
Comment (ccurran01) [inline src/main/java/com/kainos/ea/model/Capabilities.java:9]: Should Java variable names follow snakeCase pattern?

File: src/main/java/com/kainos/ea/controller/CapabilityController.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.sql.SQLException;
+
+@Api("API Hack Street Boys")
+@Path("/api")
+
```
Comment (ccurran01) [inline src/main/java/com/kainos/ea/controller/CapabilityController.java:18]: Do we need this space?

File: src/test/java/com/kainos/ea/controller/CapabilityControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import java.sql.SQLException;
+
+@ExtendWith(DropwizardExtensionsSupport.class)
+class CapabilityControllerTest {
+
+    CapabilitiesService capabilitiesService = Mockito.mock(CapabilitiesService.class);
+
+    CapabilityController capabilityController = new CapabilityController(capabilitiesService);
+
+
```
Comment (ccurran01) [inline src/test/java/com/kainos/ea/controller/CapabilityControllerTest.java:25]: Some extra blank space?

**Review by Patrick-Mc-Grath — APPROVED** [review]
Looks good

**Review by ccurran01 — APPROVED** [review]
Nice work

**Pull request merged**